### PR TITLE
planner,types: refactor the PR #8619 by using parser/types/FieldType.Clone()

### PR DESCRIPTION
**Pull request by dbjoa**
### What problem does this PR solve? 
Refactor the merged PR #8619 

### What is changed and how it works?
* Replace ``types.CloneFieldType()" by introducing the member function ``FieldType.Clone()`` in ``parser/types``

### Check List 

Tests 
 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - None

Related changes

 - ``parser/types/field_type.go``

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8641)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -85,3 +85,5 @@ require (
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20180531100431-4c381bd170b4
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
+
+replace github.com/pingcap/parser => github.com/cloud-pi/parser v0.0.0-20181211095502-1e4340922693
```
Comment (zz-jason) [inline go.mod]: irrelevant change?

Comment (winoros) [inline go.mod]: waiting for parser's pr merged.

**Comment by ngaut** [comment]
Please fix CI

**Comment by zhouqiang-cl** [comment]
It seems build failed 
@dbjoa 
```
CGO_ENABLED=0 GO111MODULE=on go build   -ldflags '-X "github.com/pingcap/parser/mysql.TiDBReleaseVersion=v2.1.0-rc.3-308-g9a9165a" -X "github.com/pingcap/tidb/util/printer.TiDBBuildTS=2018-12-11 08:45:02" -X "github.com/pingcap/tidb/util/printer.TiDBGitHash=9a9165a85db56390e5db257e82cbb45b28d78f41" -X "github.com/pingcap/tidb/util/printer.TiDBGitBranch=HEAD" -X "github.com/pingcap/tidb/util/printer.GoVersion=go version go1.11.2 linux/amd64" -X "github.com/pingcap/tidb/config.checkBeforeDropLDFlag=1"' -o bin/tidb-server tidb-server/main.go

# github.com/pingcap/tidb/ddl

ddl/ddl_api.go:1157:8: undefined: ast.AlterTableTruncatePartition

ddl/ddl_api.go:1499:15: undefined: model.ActionTruncateTablePartition

ddl/ddl_worker.go:288:137: undefined: model.ActionTruncateTablePartition

ddl/ddl_worker.go:475:7: undefined: model.ActionTruncateTablePartition

ddl/delete_range.go:246:39: undefined: model.ActionTruncateTablePartition

make: *** [server] Error 2
```

**Pull request merged**